File: 2020 年 (25 歳) 設計、デザインパターン/design-patterns-practice/Java 言語で学ぶデザインパターン入門/10 Strategy 一 アルゴリズムをごっそり切り替える/02 C++/01 old/Hand.cpp
#include "Hand.h"


Hand::Hand(int handvalue) :
    handvalue(handvalue)
{
}

Hand Hand::getHand(int handvalue)
{
    return hands[handvalue];
}

bool Hand::isStrongerThan(Hand h)
{
    return fight(h) == 1;
}

bool Hand::isWeakerThan(Hand h)
{
    return fight(h) == -1;
}


int Hand::fight(Hand h)
{
    if (this->handvalue == h.handvalue)
        return 0;
    else if ((this->handvalue + 1) % 3 == h.handvalue)
        return 1;
    else
        return -1;
}


std::string Hand::toString()
{
    return names[handvalue];
}

std::array<Hand, 3> Hand::hands
{
    {
        Hand(HANDVALUE_GUU),
        Hand(HANDVALUE_CHO),
        Hand(HANDVALUE_PAA)
    }
};

std::array<std::string, 3> Hand::names
{
    "グー",
    "チョキ",
    "パー"
};

File: 2020 年 (25 歳) 設計、デザインパターン/design-patterns-practice/Java 言語で学ぶデザインパターン入門/10 Strategy 一 アルゴリズムをごっそり切り替える/01 Java/01/06/Hand.java
public class Hand
{
    public static final int HANDVALUE_GUU = 0;
    public static final int HANDVALUE_CHO = 1;
    public static final int HANDVALUE_PAA = 2;

    public static final Hand[] hands =
	{
        new Hand(HANDVALUE_GUU),
        new Hand(HANDVALUE_CHO),
        new Hand(HANDVALUE_PAA)
    };

    public static final String[] names =
	{
        "グー",
        "チョキ",
        "パー"
    };



    // private
    // public
    private int handValue;



    private Hand(int handValue)
	{
        this.handValue = handValue;
    }

    public static Hand getHand(int handValue)
	{
        return hands[handValue];
    }

    public boolean isStrongerThan(Hand h)
	{
        // this が h より強いとき true
        return fight(h) == 1;
    }

    public boolean isWeakerThan(Hand h)
	{
        // this が h より弱いとき true
        return fight(h) == -1;
    }



    public String toString()
	{
        return names[handValue];
    }



    private int fight(Hand h)
	{
        if (this.handValue == h.handValue)
		{
			// 引き分け
            return 0;
        }
		else if ((this.handValue + 1) % 3 == h.handValue)
		{
			// this の勝ち
            return 1;
        }
		else
		{
			// this の負け
            return -1;
        }
    }
}
